crypto/curve25519/curve25519.c

```diff
@@ -31,33 +31,223 @@
+  const uint8_t point[32]) {
+  abort();
+}
+void curve25519_x25519_byte_alt(uint8_t res[32], const uint8_t scalar[32],
+  const uint8_t point[32]) {
+  abort();
+}
+void curve25519_x25519base_byte(uint8_t res[32], const uint8_t scalar[32]) {
+  abort();
 }
+void curve25519_x25519base_byte_alt(uint8_t res[32], const uint8_t scalar[32]) {
+  abort();
+}
+
+#endif // !defined(CURVE25519_ASM_CAPABLE)
+
+void x25519_NEON(uint8_t out[32], const uint8_t scalar[32],
+                 const uint8_t point[32]);
 
+#if !defined(BORINGSSL_X25519_NEON_CAPABLE)
 void x25519_NEON(uint8_t out[32], const uint8_t scalar[32],
                  const uint8_t point[32]) {
   abort();
 }
+#endif // !defined(BORINGSSL_X25519_NEON)
 
+// Run-time detection for each implementation
+
+OPENSSL_INLINE int x25519_s2n_bignum_alt_capable(void);
+OPENSSL_INLINE int x25519_s2n_bignum_no_alt_capable(void);
+
+// For aarch64, |x25519_s2n_bignum_alt_capable| returns 1 if we catagorize the
+// CPU as a CPU having a wide multiplier (i.e. "higher" throughput). CPUs with
+// this feature are e.g.: AWS Graviton 3 and Apple M1. Return 0 otherwise, so we
+// don't match CPUs without wide multipliers.
+//
+// For x86_64, |x25519_s2n_bignum_alt_capable| always returns 1. If x25519
+// s2n-bignum capable, the x86_64 s2n-bignum-alt version should be supported on
+// pretty much any x86_64 CPU.
+//
+// For all other architectures, return 0.
+OPENSSL_INLINE int x25519_s2n_bignum_alt_capable(void) {
+#if defined(OPENSSL_X86_64)
+  return 1;
+#elif defined(OPENSSL_AARCH64)
+  if (CRYPTO_is_ARMv8_wide_multiplier_capable() == 1) {
+    return 1;
+  } else {
+    return 0;
+  }
+#else
+  return 0;
 #endif
+}
+
+// For aarch64, |x25519_s2n_bignum_alt_capable| always returns 1. If x25519
+// s2n-bignum capable, the Armv8 s2n-bignum-alt version should be supported on
+// pretty much any Armv8 CPU.
+//
+// For x86_64, |x25519_s2n_bignum_alt_capable| returns 1 if we detect support
+// for bmi+adx instruction sets. Return 0 otherwise.
+//
+// For all other architectures, return 0.
+OPENSSL_INLINE int x25519_s2n_bignum_no_alt_capable(void) {
+#if defined(OPENSSL_X86_64)
+  if (CRYPTO_is_BMI2_capable() == 1 && CRYPTO_is_ADX_capable() == 1) {
+    return 1;
+  } else {
+    return 0;
+  }
+#elif defined(OPENSSL_AARCH64)
+  return 1;
+#else
+  return 0;
+#endif
+}
+
+
+// Below is the decision logic for which assembly backend implementation
+// of x25519 s2n-bignum we should use if x25519 s2n-bignum capable. Currently,
+// we support the following implementations.
+//
+// x86_64:
+//   - s2n-bignum-no-alt: hardware implementation using bmi2+adx instruction sets
+//   - s2n-bignum-alt: hardware implementation using standard instructions
+//
+// aarch64:
+//   - s2n-bignum-no-alt: hardware implementation for "small" multiplier throughput
+//   - s2n-bignum-alt: hardware implementation for "higher" multiplier throughput
+//
+// Through experiments we have found that:
```

Comment:
Did you considered determining this at runtime? Could customers call a power on benchmark to do the local experiment and figure out for their CPU the best implementation? This is out of scope for this change, I'm just thinking about cases were our experiment wouldn't produce the best results for customers.

Reply:
I guess that is possible to do, at least in theory :)